jazz2: init at 1.8.0

surfaceflinger opened the pull request:
###### Description of changes

Added a derivation for [Jazz² Resurrection](https://github.com/deathkiller/jazz2-native).

I've set few cmake flags so it won't download prebuilt dependencies, cmake also couldn't find 2 libraries so I fixed it up there too.

If you want to play, you will need original assets, preferably from [GOG](https://www.gog.com/en/game/jazz_jackrabbit_2_collection). GOG's installer sadly has to be ran with wine.

In `~/.local/share/Jazz² Resurrection` you should have this structure of all assets:
```
.
└── Source
    ├── HTML
    ├── JCSHelp
    ├── Tiles
    └── UserLevels
```

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

surfaceflinger committed:
maintainers: add surfaceflinger

RealTirex reviewed (commented):
At first glance it looks good

surfaceflinger commented:
For now I changed it to a draft from my phone so it doesn't get merged. I talked to developer and we came to conclusion that package should be called just `jazz2`. I'll update the PR once I get back home.

surfaceflinger commented:
Should be ready for review. It now loads assets from `/run/current-system` so it won't "just work" with Nix installed on other distros/macOS. In the future I'll try to spin up a macOS VM and try to get it working with nix-darwin because it apparently creates `/run/current-system` :)

If someone knows a better way without using `/run/current-system` then please tell me.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-ready-for-review/3032/1758

SuperSandro2000 commented on pkgs/games/jazz2/default.nix:
```diff
@@ -0,0 +1,46 @@
+    "-DGLFW_INCLUDE_DIR=${glfw}/include/GLFW"
+    "-DLIBOPENMPT_INCLUDE_DIR=${libopenmpt.dev}/include/libopenmpt"
+    "-DNCINE_LINUX_PACKAGE=${pname}"
+    "-DNCINE_OVERRIDE_CONTENT_PATH=/run/current-system/sw/share/${pname}/Content"
```
That will not work on systems that just use the nix package manager. Not sure what other options we have to solve this.

surfaceflinger replied:
I guess patching the source would be a solution but I doubt in myself that I'll be able to do that.

btw: This is only for autodetecting assets that come with the reimplementation. If it can't detect these files then it will check current working directory, so user can copy assets from /nix/store and let's not forget that they need to obtain and manually install assets from the original game. Check edits on PR because before I contacted the upstream I tried to explain it there.

msm-code replied:
Can you explain what it does, and what needs to be patched in order to fix this (maybe also link to upstream discussion/docs about this)?

I can help with patching if there's no better solution.

surfaceflinger replied:
https://github.com/deathkiller/jazz2-native/blob/master/Sources/Jazz2/ContentResolver.cpp#L104

jazz2 needs not only assets from original game but also has assets for its own features which can be also loaded from current directory when resolving `NCINE_OVERRIDE_CONTENT_PATH` fails

SuperSandro2000 commented on pkgs/games/jazz2/default.nix:
```diff
@@ -0,0 +1,46 @@
+
+  buildInputs = [ libGL SDL2 zlib glew glfw openal libvorbis libopenmpt ];
+  nativeBuildInputs = [ cmake ];
+  propagatedBuildInputs = [ xorg.libSM xorg.libICE xorg.libXext ];
```
Why do we need to propagate those?

surfaceflinger replied:
It crashed without xorg.libSM so before adding only this one and rebuilding - I checked executable with ldd and it was showing that these 3 were missing

msm-code commented:
Result of `nixpkgs-review pr 212426` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>jazz2</li>
  </ul>
</details>

msm-code commented:
It builds, but doesn't work for me with nixpkgs-review:

```
[nix-shell:~/.cache/nixpkgs-review/pr-212426]$ ./results/jazz2/bin/jazz2
void nCine::StandardFile::OpenStream(nCine::FileAccessMode) -> Cannot open the file "/home/msm/.config/Jazz² Resurrection/Jazz2.config"
void nCine::StandardFile::OpenStream(nCine::FileAccessMode) -> Cannot open the file "Content/Translations/en-us.mo"
void nCine::StandardFile::OpenStream(nCine::FileAccessMode) -> Cannot open the file "Cache/Translations/en-us.mo"
void nCine::MainApplication::init(std::unique_ptr<nCine::IAppEventHandler> (*)(), int, char**) -> IAppEventHandler::OnPreInit() invoked
void nCine::JoyMapping::init(const nCine::IInputManager*) -> Parsed 349 strings for 349 mappings
void nCine::Application::initCommon() -> Jazz² Resurrection v1.6.0 initializing...
void nCine::GfxCapabilities::logGLInfo() const -> --- OpenGL device info ---
void nCine::GfxCapabilities::logGLInfo() const -> Vendor: Intel
void nCine::GfxCapabilities::logGLInfo() const -> Renderer: Mesa Intel(R) UHD Graphics (CML GT2)
void nCine::GfxCapabilities::logGLInfo() const -> OpenGL Version: 4.6 (Core Profile) Mesa 22.0.4
void nCine::GfxCapabilities::logGLInfo() const -> GLSL Version: 4.60
void nCine::GfxCapabilities::logGLCaps() const -> --- OpenGL device capabilities ---
void nCine::GfxCapabilities::logGLCaps() const -> GL_MAX_TEXTURE_SIZE: 16384
void nCine::GfxCapabilities::logGLCaps() const -> GL_MAX_TEXTURE_IMAGE_UNITS: 32
void nCine::GfxCapabilities::logGLCaps() const -> GL_MAX_UNIFORM_BLOCK_SIZE: 65536
void nCine::GfxCapabilities::logGLCaps() const -> GL_MAX_UNIFORM_BUFFER_BINDINGS: 90
void nCine::GfxCapabilities::logGLCaps() const -> GL_MAX_VERTEX_UNIFORM_BLOCKS: 15
void nCine::GfxCapabilities::logGLCaps() const -> GL_MAX_FRAGMENT_UNIFORM_BLOCKS: 15
void nCine::GfxCapabilities::logGLCaps() const -> GL_UNIFORM_BUFFER_OFFSET_ALIGNMENT: 32
void nCine::GfxCapabilities::logGLCaps() const -> GL_MAX_VERTEX_ATTRIB_STRIDE: 2048
void nCine::GfxCapabilities::logGLCaps() const -> GL_MAX_COLOR_ATTACHMENTS: 8
void nCine::GfxCapabilities::logGLCaps() const -> ---
void nCine::GfxCapabilities::logGLCaps() const -> GL_KHR_debug: 1
void nCine::GfxCapabilities::logGLCaps() const -> GL_ARB_texture_storage: 1
void nCine::GfxCapabilities::logGLCaps() const -> GL_EXT_texture_compression_s3tc: 1
void nCine::GfxCapabilities::logGLCaps() const -> GL_OES_compressed_ETC1_RGB8_texture: 0
void nCine::GfxCapabilities::logGLCaps() const -> GL_AMD_compressed_ATC_texture: 0
void nCine::GfxCapabilities::logGLCaps() const -> GL_IMG_texture_compression_pvrtc: 0
void nCine::GfxCapabilities::logGLCaps() const -> GL_KHR_texture_compression_astc_ldr: 1
void nCine::GfxCapabilities::logGLCaps() const -> ---
static void nCine::RenderResources::create() -> Creating rendering resources...
static void nCine::RenderResources::create() -> Rendering resources created
void nCine::Application::initCommon() -> Application initialized
void nCine::StandardFile::OpenStream(nCine::FileAccessMode) -> Cannot open the file "Cache/Animations/cache.index"
void GameEventHandler::RefreshCache() -> Cannot open "./Source/Anims.j2a" file! Make sure Jazz Jackrabbit 2 files are present in "Source/" directory.
virtual void GameEventHandler::OnInit() -> Rendering resolution: 720x405
void nCine::Application::initCommon() -> IAppEventHandler::OnInit() invoked
virtual void GameEventHandler::OnResizeWindow(int, int) -> Rendering resolution: 1920x1037
void nCine::StandardFile::OpenStream(nCine::FileAccessMode) -> Cannot open the file "Cache/Tilesets/easter99.j2t"
void nCine::StandardFile::OpenStream(nCine::FileAccessMode) -> Cannot open the file "Cache/Tilesets/carrot1.j2t"
void nCine::StandardFile::OpenStream(nCine::FileAccessMode) -> Cannot open the file "Cache/Tilesets/diam2.j2t"
void nCine::StandardFile::OpenStream(nCine::FileAccessMode) -> Cannot open the file "Content/Metadata/UI/MainMenu.res"
void nCine::StandardFile::OpenStream(nCine::FileAccessMode) -> Cannot open the file "Content/Animations/UI/font_small.png.font"
void nCine::StandardFile::OpenStream(nCine::FileAccessMode) -> Cannot open the file "Content/Animations/UI/font_medium.png.font"
Segmentation fault (core dumped)
```

This is probably related to the `/var/run/current-system` issue?

surfaceflinger committed:
jazz2: init at 1.8.0

surfaceflinger commented:
I've rebased on master, updated to 1.7.0 and content is now in a separate derivation called `jazz2-content` which allows `jazz2` to no longer depend on current-system and instead uses nix store. I also got rid of propagated build inputs because it looks like these simply weren't needed. Please review, I think that this can get merged now.

SuperSandro2000 commented on pkgs/games/jazz2/game.nix:
```diff
@@ -0,0 +1,42 @@
+, zlib
+}:
+
+let
+  sources = (import ./sources.nix) { inherit fetchFromGitHub; };
+in
+stdenv.mkDerivation rec {
+  inherit (sources) pname version src;
```
```suggestion
stdenv.mkDerivation rec {
  pname = "jazz2";
  version = "1.7.0";

  src = fetchFromGitHub {
    owner = "deathkiller";
    repo = "jazz2-native";
    rev = version;
    sha256 = "U+V3332z09085p+i5Nev2dBiAFH6FApXgZS/XJe6ijQ=";
  };
```

SuperSandro2000 commented on pkgs/games/jazz2/sources.nix:
```diff
@@ -0,0 +1,16 @@
+{ fetchFromGitHub }:
+
+let
+  pname = "jazz2";
+  version = "1.7.0";
+in
+{
+  inherit version pname;
+
+  src = fetchFromGitHub {
+    owner = "deathkiller";
+    repo = "${pname}-native";
+    rev = version;
+    sha256 = "U+V3332z09085p+i5Nev2dBiAFH6FApXgZS/XJe6ijQ=";
+  };
+}
```
```suggestion
```

SuperSandro2000 commented on pkgs/games/jazz2/content.nix:
```diff
@@ -0,0 +1,28 @@
+{ fetchFromGitHub
+, lib
+, stdenv
+}:
+
+let
+  sources = (import ./sources.nix) { inherit fetchFromGitHub; };
+in
+stdenv.mkDerivation rec {
+  pname = "${sources.pname}-content";
+
+  inherit (sources) version src;
+
+  dontBuild = true;
+
+  installPhase = ''
+    cp -r Content $out
+  '';
```
```suggestion
{ lib
, jazz2
, stdenv
}:

stdenv.mkDerivation rec {
  pname = "jazz2-content";
  inherit (jazz2) version src;

  dontBuild = true;

  installPhase = ''
    mv Content $out
  '';
```

SuperSandro2000 reviewed (commented):
we can make this a bit nicer

surfaceflinger commented:
That's so cool, I'll apply and update later because of easter (:

SuperSandro2000 commented on pkgs/games/jazz2/game.nix:
```diff
@@ -0,0 +1,48 @@
+}:
+
+stdenv.mkDerivation rec {
+
```
```suggestion
```

SuperSandro2000 commented on pkgs/games/jazz2/content.nix:
```diff
@@ -0,0 +1,25 @@
+, stdenv
+}:
+
+stdenv.mkDerivation rec {
+  pname = "jazz2-content";
+
+  inherit (jazz2) version src;
+
+  dontBuild = true;
+
+  installPhase = ''
+    cp -r Content $out
+  '';
```
We could just use runCommand here

surfaceflinger replied:
I'm not really sure how I'm supposed to do that 😕

SuperSandro2000 commented on pkgs/games/jazz2/game.nix:
```diff
@@ -0,0 +1,48 @@
+    homepage = "https://github.com/deathkiller/jazz2-native";
+    license = licenses.gpl3;
+    maintainers = with maintainers; [ surfaceflinger ];
+    platforms = with platforms; linux;
```
```suggestion
    platforms = platforms.linux;
```

surfaceflinger commented:
I think it's fine now? Not sure if I used runCommand properly though :)
